### DISCUSSION TITLE
Migrating to Clap

### DISCUSSION THREAD

**Discussion by gierens** [eza-community/eza#195] [Development]
So far the builtin argparser has done it's job well I'd say or at least hasn't prevented us from implementing any features as far as I'm aware. With some fuzzy searching through the code it's also fairly easy to understand how to add a new flag and such. That's why switching to something more mainstream hasn't been a huge priority thus far.

Having said that I do however agree with those raising voices that in the long run it might be beneficial to move to clap since it's sort of the standard nowadays, well understood and documented and very powerful especially with the macros. This will require some heavy re-engineering though, and we got to be careful not to produce a mess but a very clean and easy to review PR. Just looking at `src/options` I feel like this might require more than one person. Therefore let's discuss the details here.

**Comment by daviessm** [comment]
I'd be happy to help. So the first discussion point would be - derive or builder?

**Reply by gierens** [reply]
If possible I would go with derive, it's just so much easier to understand when reading the code in my opinion.

**Reply by cafkafk** [reply]
Personally, I prefer builder somewhat. But I have no good reason beyond just liking it more. I'm fine with derive thou.

**Reply by MartinFillon** [reply]
Derived looks easier to implement and to read

**Reply by PThorpe92** [reply]
I've always gone with builder myself, but either way it's a major convenience upgrade for the code base. I feel like whoever makes the first solid PR can choose lmao.. it's not going to be simple from the looks of it.

**Reply by MartinFillon** [reply]
There is one small issue that I just encountered, we can't ovewrite -h option as they reserve it for help but we need it for --header option

**Reply by PThorpe92** [reply]
That actually is probably more of an issue than it may seem at first... Myself, and probably many many others have the -h flag in my alias's on both my computers. Breaking that would be pretty tragic and probably upset a few people.

Looking into this

**Comment by MartinFillon** [comment]
https://github.com/eza-community/eza/pull/197 -- POC of the parser

DO NOT MERGE IT

**Comment by sbatial** [comment]
This would also make #37 much easier/possible at all I think which would be nice!
